govcd/saml_auth.go

```diff
@@ -0,0 +1,292 @@
+// getSamlTokenRequestBody. This request is submited to ADFS server endpoint
+// "/adfs/services/trust/13/usernamemixed" and `RequestedSecurityTokenTxt` is expected in response
+func vcdAuthorizeSamlGetSamlAuthToken(vcdCli *VCDClient, user, pass, samlEntityId, authEndpoint, org string) (string, error) {
+	requestBody := getSamlTokenRequestBody(user, pass, samlEntityId, authEndpoint)
+	samlTokenRequestBody := strings.NewReader(requestBody)
+	tokenRequestResponse := types.AdfsAuthResponseEnvelope{}
+
+	// Post to ADFS endpoint "/adfs/services/trust/13/usernamemixed"
+	authEndpointUrl, err := url.Parse(authEndpoint)
+	if err != nil {
+		return "", fmt.Errorf("SAML - error parsing authentication endpoint %s: %s", authEndpoint, err)
+	}
+	req := vcdCli.Client.NewRequest(nil, http.MethodPost, *authEndpointUrl, samlTokenRequestBody)
+	req.Header.Add("Content-Type", types.SoapXML)
+	resp, err := vcdCli.Client.Http.Do(req)
+	resp, err = checkRespWithErrType(resp, err, &types.AdfsAuthErrorEnvelope{})
+	if err != nil {
+		return "", fmt.Errorf("SAML - ADFS token request query failed for RPT ID ('%s'): %s",
+			samlEntityId, err)
+	}
+
+	err = decodeBody(resp, &tokenRequestResponse)
+	if err != nil {
+		return "", fmt.Errorf("SAML - error decoding ADFS token request response: %s", err)
+	}
+
+	tokenString := tokenRequestResponse.Body.RequestSecurityTokenResponseCollection.RequestSecurityTokenResponse.RequestedSecurityTokenTxt.Text
+
+	return tokenString, nil
+}
+
+// vcdAuthorizeSignToken submits a SIGN token received from ADFS server and gets regular vCD
+// "X-Vcloud-Authorization" token in exchange
+func vcdAuthorizeSignToken(vcdCli *VCDClient, base64GzippedSignToken, org string) (string, error) {
+	url, err := url.Parse(vcdCli.Client.VCDHREF.Scheme + "://" + vcdCli.Client.VCDHREF.Host + "/api/sessions")
+	if err != nil {
+		return "", fmt.Errorf("SAML error - could not parse URL for posting SIGN token: %s", err)
+	}
+
+	signHeader := http.Header{}
+	signHeader.Add("Authorization", `SIGN token="`+base64GzippedSignToken+`",org="`+org+`"`)
+
+	req := vcdCli.Client.newRequest(nil, nil, http.MethodPost, *url, nil, vcdCli.Client.APIVersion, signHeader)
+	resp, err := checkResp(vcdCli.Client.Http.Do(req))
+	if err != nil {
+		return "", fmt.Errorf("SAML - error submitting SIGN token for authentication to %s: %s", req.URL.String(), err)
+	}
+	err = decodeBody(resp, nil)
+	if err != nil {
+		return "", fmt.Errorf("SAML - error decoding body SIGN token auth response: %s", err)
+	}
+
+	accessToken := resp.Header.Get("X-Vcloud-Authorization")
+	util.Logger.Printf("[DEBUG] SAML - setting access token for further requests")
+	return accessToken, nil
+}
+
+// getSamlTokenRequestBody returns a SAML Token request body which is accepted by ADFS server
+// endpoint "/adfs/services/trust/13/usernamemixed".
+// The payload is not configured as a struct and unmarshalled because Go's unmarshalling changes
+// structure so that ADFS does not accept the payload
```

Comment:
Sadly there is no "authoritative" link. Microsoft usually put C# SDK examples instead of plain body. There are forums like https://techdebug.com/blog/2015/02/07/saml-assertion-adfs2-via-soap-endpoint/ but that doesn't look too authoritative to put into code.

Reply:
so is this from live logs?

Reply:
In one form or another I guess - yes. There are plenty places where people share this payload/do postman examples, etc.

Reply:
oh. So it's "randomly" shared in internet :)